src/server/dflycmd.cc

```diff
@@ -351,24 +408,19 @@ void DflyCmd::StopFullSyncInThread(FlowInfo* flow, EngineShard* shard) {
 }
 
 OpStatus DflyCmd::StartStableSyncInThread(FlowInfo* flow, Context* cntx, EngineShard* shard) {
-  // Register journal listener and cleanup.
-  uint32_t cb_id = 0;
-  JournalWriter* writer = nullptr;
+  // Create streamer for shard flows.
+  JournalStreamer* streamer = nullptr;
   if (shard != nullptr) {
-    writer = new JournalWriter{};
-    auto journal_cb = [flow, cntx, writer](const journal::Entry& je) mutable {
-      writer->Write(je);
-      if (auto ec = writer->Flush(flow->conn->socket()); ec)
-        cntx->ReportError(ec);
-    };
-    cb_id = sf_->journal()->RegisterOnChange(std::move(journal_cb));
+    streamer = new JournalStreamer{sf_->journal(), cntx};
+    streamer->Start(flow->conn->socket());
   }
 
-  flow->cleanup = [this, cb_id, writer, flow]() {
-    if (writer)
-      delete writer;
-    if (cb_id)
-      sf_->journal()->UnregisterOnChange(cb_id);
+  // Register cleanup.
+  flow->cleanup = [this, streamer, flow]() {
```

Comment:
Is it possible that flow cleanup will be triggered between the start above and the update of the cleanup lambda?

Reply:
No, because unlike the replica it has a per-session mutex that is locked on state transitions

Reply:
And it doesn't contain any await points